### PR TITLE
Set the default HTTP retries to 5

### PR DESCRIPTION
**Pull request by io7m**
## Description

Currently, the default HTTP retries are 0 if not explicitly set in code. This leads to us having to track down lots of different failures when a server inevitably times out during an OPDS/ODL import or similar.

## Motivation and Context

Affects: https://www.notion.so/lyrasis/ODL2-import-monitor-crashed-on-gorgon-4771eb7a31144e0d9739431d2fc4cfdd

## How Has This Been Tested?

I ran the test suite.

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/util/http.py [hunk 1]
```diff
@@ -255,15 +255,14 @@ def _request_with_timeout(
...
-                        #  Are there some that we want to ensure get retried (e.g., 502, which we get a lot).
-                        retry_strategy = Retry(total=max_retry_count)
-                        adapter = HTTPAdapter(max_retries=retry_strategy)
-
-                        session.mount("http://", adapter)
-                        session.mount("https://", adapter)
+                    # TODO: Verify the status codes for which it retries.
+                    #  Are there any status codes for which we would NOT want to retry (e.g., 401, 403)?
+                    #  Are there some that we want to ensure get retried (e.g., 502, which we get a lot).
+                    retry_strategy = Retry(total=max_retry_count)
```
[the remark this answers is not in the document]
Comment (io7m) [reply core/util/http.py:261]: Reading through the sources suggested it would, but I'll add a test.

[the remark this answers is not in the document]
Comment (io7m) [reply core/util/http.py:261]: Testing has uncovered multiple problems. :tada:

[the remark this answers is not in the document]
Comment (io7m) [reply core/util/http.py:261]: Amusingly, it seems like the retry mechanism might have been broken this entire time. There's a nice big `TODO` and `XXX` comment above the code that's supposed to handle retries that essentially says "Verify if this works...". Shockingly, it doesn't.

I've managed to get it to work though, just waiting for the test suite to complete.

[the remark this answers is not in the document]
Comment (io7m) [reply core/util/http.py:261]: I'm having to refactor `http.py` somewhat. It's a mess.

[the remark this answers is not in the document]
Comment (io7m) [reply core/util/http.py]: Ugh, now there's a test reliably failing in CI that's not failing locally. Investigating... :hammer:

**Comment by io7m** [comment]
I'm not sure I agree with that coverage report. We have more tests now than we had before, testing code that wasn't tested...

**Comment by io7m** [comment]
This is ready to go. It highlighted the fact that code that was expecting retries wasn't getting them.

**Comment by jonathangreen** [comment]
@io7m one case where we don't want this behavior, that I didn't think of before now, is in our authentication providers. If there is a problem there, I we want them to fail quickly, rather then retrying. 

I believe some of them will hit this code path and start using retries via this pull request. `MilleniumPatronAPI` is the big one I can think of, but it would be good to look at all the classes that inherit from `BasicAuthenticationProvider`, and set their retries to 0.

**Comment by io7m** [comment]
>  it would be good to look at all the classes that inherit from `BasicAuthenticationProvider`, and set their retries to 0.

Taking a look!

**Comment by io7m** [comment]
I have found a few authentication providers that need to opt-out of retries. I'm working on it now.

Amusingly, some of the authentication providers are exempt because they don't go through `http.py`. What's better than an inconsistent API? An inconsistent API that's inconsistently used! :tada:

**Comment by io7m** [comment]
I don't think there's anything more to do here.

File: core/util/http.py [hunk 2]
```diff
@@ -191,6 +191,9 @@ def request_with_timeout(
             url, sessions.Session.request, http_method, *args, **kwargs
         )
 
+    # The set of status codes on which a retry will be attempted (if the number of retries requested is non-zero).
+    RETRY_STATUS_CODES = list(range(500, 600))
```
Comment (jonathangreen) [inline core/util/http.py:195]: Minor: This is fine, but we can probably be a bit more precise with it. 

I'd suggest: 
```python
RETRY_STATUS_CODES = [429, 500, 502, 503, 504]
```

**Review by jonathangreen — APPROVED** [review]
This looks good to me. I have a couple minor suggestions for improvements, but feel free to merge as is or add them and merge.

File: core/util/http.py [hunk 3]
```diff
@@ -255,15 +258,13 @@ def _request_with_timeout(
...
             if make_request_with == sessions.Session.request:
                 with sessions.Session() as session:
-                    if max_retry_count is not None:
-                        # TODO: Verify the status codes for which it retries.
-                        #  Are there any status codes for which we would NOT want to retry (e.g., 401, 403)?
-                        #  Are there some that we want to ensure get retried (e.g., 502, which we get a lot).
-                        retry_strategy = Retry(total=max_retry_count)
-                        adapter = HTTPAdapter(max_retries=retry_strategy)
+                    retry_strategy = Retry(
+                        total=max_retry_count, status_forcelist=cls.RETRY_STATUS_CODES
```
[the remark this answers is not in the document]
Comment (io7m) [reply core/util/http.py:262]: I'd have to search again, but I think at least somewhere in the code was requesting a backoff factor (and clearly not ever getting one, as it wasn't getting retries in the first place).

**Comment by pcustic** [comment]
@io7m Just one note - this PR had failing lint tests that got merged.

**Pull request merged**